### PR TITLE
feat: allow sorting of articles

### PR DESCRIPTION
**Pull request by scmmishra**
## Reordering from the UI 

https://www.loom.com/share/cea10e3e08f84d00b6a1e9d3af0bcc05

## Creating a new article and/or moving articles between categories

https://www.loom.com/share/e327203a064f4f32a0ae2e3bc533ecbc

### CODE REVIEW

File: app/javascript/dashboard/routes/dashboard/helpcenter/components/ArticleTable.vue [hunk 1]
```diff
@@ -89,4 +148,9 @@ export default {
     border: 0;
   }
 }
+
+.article-ghost-class {
+  opacity: 0.5;
+  background-color: var(--s-50);
```
Comment (nithindavid) [inline app/javascript/dashboard/routes/dashboard/helpcenter/components/ArticleTable.vue:154]:
```suggestion
  background-color: var(--s-50);
  border-bottom: 1px solid var(--w-300);
```

Can you please see if the border color gives more  visual cue about the position it's getting inserted ?

**Review by nithindavid — APPROVED** [review]
Tested on local FE Part works as expected

**Comment by nithindavid** [comment]
@scmmishra Any cue to let the user know we can re-order articles? 
We could add `    cursor: grab;` in `app/javascript/dashboard/routes/dashboard/helpcenter/components/ArticleItem.vue:141`

**Comment by scmmishra** [comment]
@nithindavid yeah, I've been thinking of the UX part too. 

> We could add `cursor: grab;`

this looks like a good idea. 


I also though adding a drag-handle at the start of the list item, wdyt of that?

File: app/javascript/dashboard/routes/dashboard/helpcenter/components/ArticleTable.vue (hunk 1 above)
Comment (scmmishra) [reply app/javascript/dashboard/routes/dashboard/helpcenter/components/ArticleTable.vue:154]: Sure, we could do more improvements to this. But the table format is limiting. For instance, the `<tr>` cannot have left and right padding. If we could make that a grid/flex, we can make the styling much better suited for a dragging case

File: app/models/article.rb [hunk 2]
```diff
@@ -115,6 +120,38 @@ def increment_view_count
...
   private
 
+  def category_id_changed_action
+    # We need to update the position of the article in the new category
+    return unless persisted?
+
+    # this means the article is just created
+    # and the category_id is newly set
+    # and the position is already present
+    return if created_at_before_last_save.nil? && position.present? && category_id_before_last_save.nil?
```
Comment (scmmishra) [inline app/models/article.rb:130]: @tejaswinichile @sojan-official need your thoughts on this logic.. This is to ensure the category ID is not overridden on create, if the position is supplied when creating the article

**Comment by nithindavid** [comment]
@muhsin-k
The icons can be improved with fluent suggested ones

1. Sort 
image

2. Filter
image
cc: @scmmishra

File: app/controllers/api/v1/accounts/articles_controller.rb [hunk 3]
```diff
@@ -68,4 +78,12 @@ def list_params
   def set_current_page
     @current_page = params[:page] || 1
   end
+
+  def update_positions(positions_hash)
```
Comment (tejaswinichile) [inline app/controllers/api/v1/accounts/articles_controller.rb:82]: Can you move this method to `Article.rb`?

**Review by tejaswinichile — CHANGES_REQUESTED** [review]

File: app/controllers/api/v1/accounts/articles_controller.rb (hunk 3 above)
Comment (scmmishra) [reply app/controllers/api/v1/accounts/articles_controller.rb]: @tejaswinichile done, can you check

File: db/migrate/20230405104405_update_position_for_articles.rb [hunk 4]
```diff
@@ -0,0 +1,30 @@
+class UpdatePositionForArticles < ActiveRecord::Migration[6.1]
```
Comment (tejaswinichile) [inline db/migrate/20230405104405_update_position_for_articles.rb:1]: 💯

File: app/models/article.rb (hunk 2 above)
Comment (tejaswinichile) [reply app/models/article.rb:138]: This seems okay.

**Review by tejaswinichile — APPROVED** [review]

**Pull request merged**